### PR TITLE
Lint Python code for undefined names

### PR DESCRIPTION
**Pull request by cclauss**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1721)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -93,5 +93,8 @@ matrix:
     - language: python
       python: "3.7"
       env: TOXENV=py37
-      dist: xenial # required for Python >= 3.7
```
Comment (neuromage) [inline .travis.yml]: Is this no longer necessary?

Comment (cclauss) [reply .travis.yml]: Correct.

**Comment by ajchili** [comment]
Hey @cclauss, is there a way that we can exclude a directory with this change? The `backend/src/apiserver/visualization` directory uses dependency injection to set variables for the visualization python files and would result in these new tests always failing. I am also opened to suggestions and better approaches for dependency injection in Python.

For reference, the current approach uses Jupyter Notebook cells to inject variables into a visualization.
For instance:
```python
variables = "{}".format({"x": 2})
```
Can be accessed in a separate file like:
```python
if variables.get("x") == 2:
    pass
```

**Comment by cclauss** [comment]
Three ways to do this:
1. Put __# flake8: noqa__ at the top of each file in __backend/src/apiserver/visualization__.
2. Put __# noqa__ at the end of each line that flake8 complains about.
3. Add __--exclude=./.*,.backend/src/apiserver/visualization__ to the flake8 command.
4. Add a __.flake8__ file with a similar exclude statement at the root of the repo.

**Comment by ajchili** [comment]
Thanks for the quick response! I think adding either **# flake8: noqa** to the top of visualization files, excluding exporter.py, server.py, exporter_text.py, and server_test.py, or creating the .flake8 file with the same exclusions would suffice and prevent future test failure. Let me know what you think!

**Comment by ajchili** [comment]
+1 for global magic variable deceleration. It seems like your branch may be behind master as `roc_curve.py` is still using **input_path** rather than the **source** variable.

File: backend/src/apiserver/visualization/roc_curve.py [hunk 2]
```diff
@@ -27,12 +27,7 @@
 # The following variables are provided through dependency injection. These
 # variables come from the specified input path and arguments provided by the
 # API post request.
-#
-# is_generated
-# source
-# target_lambda
-# trueclass
-# true_score_column
+global input_path, is_generated, target_lambda, true_score_column, trueclass
```
Comment (ajchili) [inline backend/src/apiserver/visualization/roc_curve.py]: Can now be
```python
global variables, source
```

**Comment by ajchili** [comment]
It looks like using
```python
global variables, source
```
does not indicate to flake8 that the variable is declared as seen in the [travis logs](https://travis-ci.com/kubeflow/pipelines/jobs/226848818#L296). For visualization files like **roc_curve.py** and **tfdv.py** I think including them in a **.flask8** file would be the best approach. But, this does not resolve the other undeclared variables that are present in other files. @Ark-kun or @cclauss do you have any suggestions for how this PR can be merged/resolved in an appropriate amount of time while not requiring major work from @cclauss? Perhaps including **# flake8: noqa** at the top of the files that include undeclared variables and adding a TODO would suffice for this PR and each file can be addressed in a subsequent PR?

File: backend/src/apiserver/visualization/roc_curve.py [hunk 3]
```diff
@@ -34,7 +36,7 @@
 # trueclass
 # true_score_column
 
-if "is_generated" is not in variables or variables["is_generated"] is False:
+if not variables.get("is_generated"):
```
Comment (ajchili) [inline backend/src/apiserver/visualization/roc_curve.py]: I think that we should revert this change and move it to a separate PR. What are your thoughts?

File: .travis.yml [hunk 4]
```diff
@@ -94,5 +94,9 @@ matrix:
...
       script: *1
+    - name: "Lint Python code with flake8"
+      language: python
+      python: "3.7"
+      install: pip install flake8
+      script: flake8 . --count --exclude=./.*,*/visualization/tfdv.py --select=E9,F63,F7,F82 --show-source --statistics
```
Comment (ajchili) [inline .travis.yml]: For consistency can we either include **roc_curve.py** within this exclude statement or add `# flake8: noqa TODO` to **tfdv.py**.

File: backend/src/apiserver/visualization/roc_curve.py (hunk 3 above)
Comment (cclauss) [reply backend/src/apiserver/visualization/roc_curve.py]: That change has been in #1878 for 4 days (a long time for a syntax error) and was not acted on so I bundled it it here

File: .travis.yml (hunk 4 above)
Comment (cclauss) [reply .travis.yml]: __roc_curve.py__ passes the flake8 tests.  Why would we want to move it to this exclude statement?  We only moved __tfdv.py__ to this exclude statement because its pytests are overzealous in that the complain if a comment is added to the file.

Comment (ajchili) [reply .travis.yml]: Previously the tests failed before `# flake8: noqa TODO` was added to **roc_curve.py** as mentioned in [this comment](https://github.com/kubeflow/pipelines/pull/1721#issuecomment-523248857). I believe we could remove **tfdv.py** from the exclude statement and just add `# flake8: noqa TODO` at the top of the file (at least from my understanding of flask8) and mimic the implementation within **roc_curve.py**. It would be simpler to keep the way we ignore these files identical and new visualizations that are added can follow the same structure defined within this PR to enforce consistency.

Comment (cclauss) [reply .travis.yml]: If we add __# flake8: noqa TODO__ to the top of __tfdv.py__ then its pytest fails.

Comment (ajchili) [reply .travis.yml]: Ah I realize the issue now, it is due to the snapshots being out of date. We can update the snapshots if we add **# flake8: noqa TODO** to the top of **tfdv.py**.

Comment (ajchili) [reply .travis.yml]: As a follow up
https://github.com/kubeflow/pipelines/blob/94aa0788d5523de92c803407abb470ea053421f5/backend/src/apiserver/visualization/test_exporter.py#L59-L62
can be changed to use **test.py** rather than **tfdv.py** at line 61. This would prevent the overzealous unit tests from getting in the way of future changes to **tfdv.py** while also ensuring `create_cell_from_file` works as intended. After this is taken care of I think this is ready to be merged. Thanks for all your work @cclauss!

**Comment by ajchili** [comment]
Hey @cclauss, one last thing needs to be done. The unit tests for `exporter.py` need to be updated. While the the [snapshottest repository](https://github.com/syrusakbary/snapshottest#usage-with-pytest) state that this can be done by running `python3 exporter_test.py --snapshot-update` within the **backend/src/apiserver/visualization** directory, this command does not seem to work. Instead you can likely replace
https://github.com/kubeflow/pipelines/blob/3b7340f25851055c320fc391fe350727f8737fb4/backend/src/apiserver/visualization/snapshots/snap_test_exporter.py#L79-L104
with the contents of [test.py](https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/visualization/test.py) and locally run the unit test to confirm that it is now valid. Sorry to make you jump through all these hoops.

**Comment by ajchili** [comment]
/lgtm
Thanks for this change @cclauss, it'll dramatically increase code quality!
/cc @IronPan for review/approval
